src/pages/404.astro

```diff
@@ -15,14 +12,16 @@ export const prerender = false;
 
 Astro.response.status = 404;
 
-const { locale } = Astro.params as { locale: SiteLocale };
+const localeFromPath = Astro.params.locale as SiteLocale;
+const localeFromQuery = Astro.url.searchParams.get('locale') as SiteLocale;
+const locale = localeFromQuery || localeFromPath;
```

Comment:
This is now needed to have the `Astro.rewrite('/404/')` work in `[locale]/[...path]/index.astro`. It's the tradeoff for no longer needing to cast the response from the page query. Worth it? Or rather have that one fail as it did before?